draft/XXXX-multipart-upload/README.md

```diff
@@ -0,0 +1,115 @@
+settle them in their final home in the server (as is done currently, directly
+with the file contents).
+
+Once the work is completed, the server can delete the upload collection from the
+database.
+
+### API design
+
+This RFC adds an `uploads` section to the Multinet API, which contains the
+following endpoints:
+
+- `POST /api/uploads` - creates a new upload object by selecting a random,
+non-preexisting SHA string, then creating an Arango collection keyed by the
+string. This endpoint returns the ID string.
+- `POST /api/uploads/<upload-id>/chunk?sequence=<sequence-number>` - uploads one
+chunk of data to the named upload object. This endpoint returns the sequence
+number, and may raise various status codes to indicate problems.
+- `POST /api/uploads/<upload-id>/finalize` - finalizes the IDed upload by
+preventing further calls to `POST /api/uploads/<upload-id>/chunk` and performing
+a checksum on the settled data.
+- `DELETE /api/uploads/<upload-id>` - removes the database collection associated
+with the given ID.
+
+### File-like interface to upload collection
+
+A Python class named `UploadReader` which implements the file-like interface
+for reading data from an upload collection will be used by the serverside to
+interact with the data uploaded to the database using the `/api/uploads`
+endpoints. It should be constructed with a string ID referring to an upload, and
+it should implement the `seek()`, `read()`, and `readline()` methods as though
+the various documents in the collection constitute a regular file as opened by
+Python's `open()` function.
+
+### Usage
+
+Currently, file uploads are done by file type using the `uploaders` API section.
+There, a file of the specified type is sent to a specific `POST
+/api/uploaders/<file-type>` endpoint, which in turn handles the full file data
+including validation and ingestion to an Arango database and collection.
+
+Under this RFC, file upload becomes detached from file ingestion. Now, the
+correct sequence of `/api/uploads` calls must be made in order to settle the
+data in the database in an intermediate form.
+
+Next, the client can invoke one of the uploader paths, which will in turn
+instantiate an `UploadReader` object. This file-like object can be used to
+perform validation and bulk ingestion into a Multinet data table proper. The
+main advantage of this usage is that an `UploadReader` does not exhaust its data
+iterator after one pass; it can be rewound by using the `.seek()` method. This
+means that both validation and ingestion can be performed without a memory
+blowup caused by exhausting an HTTP request body iterator into a string in
+memory.
+
+Finally, the client (or the uploader methods) can invoke the `DELETE` endpoint
+to remove the intermediate upload data, leaving the server in a clean state.
+
+## Questions
+
+1. How do we detect upload interruptions? What is the right way to resume such
+   an interrupted upload?
```

Comment:
I don't really think we need to "detect" upload interruptions (from the server, that is), we just need to have a way to resume the uploads. This would necessitate another couple endpoints:

* `GET /api/uploads` - List all active uploads for the current user
* `GET /api/uploads/<upload-id>` - Get the metadata about a specific upload

So that, if the upload is interrupted or something, the client can get this list of active uploads, and show the user, so that they may manually resume this upload. The second endpoint mentioned above may not be strictly necessary for this, but may be a good idea to have.

Reply:
This feels right.

So when this happens, and the client shows the user a list of uploads, what does the user do? They select an upload, and then supply the file to upload once more?

Reply:
Yeah, that sounds right. I'm assuming the data-model for the "status" of an upload includes a byte-count or something similar, so when the user selects a file to resume uploading, the client just starts from that byte-count and resumes sending chunks.

I suppose as part of this, we should properly flesh out what data that status contains. I can think of a few fields off the top of my head:

```
{
  "id": <uuid>,
  "uploaded": <int>,       # represents bytes uploaded
  "collection": <string>,  # collection that stores the upload data
  ...
}
```

I think this would be stored in the `_system` "uploads" collection that I mentioned elsewhere.

Reply:
Ok. I'll add some text about this.

I think the model for this will be a bit more complex than simply what prefix of bytes has been uploaded. One benefit of multipart upload is that you can send several chunks in parallel. So if an "earlier" chunk fails while a "later" one succeeds, we'd need some way to encode that.

Reply:
Since we've pivoted to a parallel upload approach, the metadata that we'd want to store would be a bit different regarding upload progress. Perhaps something like:

```
{
  "id": <uuid>,
  "collection": <string>,  # Collection that stores the upload data
  "filename": <string>,    # The name of the uploading file client-side (only used as a hint to the client) 
  "size": <int>,           # The size of the total file in bytes (as reported by the client)
  "chunk_size": <int>,     # The size of each chunk in bytes
  "progress": {
    [chunk_seq: int]: true
  }
}
```

The new thing here is the `progress` field, which would basically be a checklist for each chunk sequence. That is, when a chunk is uploaded successfully, the following is performed:
```
upload_doc["progress"][chunk_seq] = true
```

This means that on the client, to resume an interrupted upload, you would begin uploading all chunks who's chunk sequence (or index) is not in the `progress` field.

This also means that determining the upload progress percentage can be done by dividing the number of entries in the `progress` field by the number of chunks (which is `size`/`chunk_size`).

Let me know what you think.